File: 1092-shortest-common-supersequence/1092-shortest-common-supersequence.py
class Solution:
    def shortestCommonSupersequence(self, str1: str, str2: str) -> str:
        """Find the longest common subsequence"""
        m=len(str1)
        n=len(str2)
        dp=[[0]*(n+1) for _ in range(m+1)]
        for i in range(1,m+1):
            for j in range(1,n+1):
                if str1[i-1]==str2[j-1]:
                    dp[i][j]=1+dp[i-1][j-1]
                else:
                    dp[i][j]=max(dp[i-1][j],dp[i][j-1])
        #lcs=dp[m][n]
        """Finding the lcs in both the strings"""
        i=m
        j=n
        ans=[]
        while i>0 and j>0:
            if str1[i-1]==str2[j-1]:
                ans.append(str1[i-1])
                i-=1
                j-=1
            else:
                if i>=1 and j>=1:
                    if dp[i-1][j]>dp[i][j-1]:
                        ans.append(str1[i-1])
                        i-=1
                    else:
                        ans.append(str2[j-1])
                        j-=1
        while j>0:
            ans.append(str2[j-1])
            j-=1
        while i>0:
            ans.append(str1[i-1])
            i-=1
        return "".join(ans[::-1]) 
        
        

File: 1092-shortest-common-supersequence/1092-shortest-common-supersequence.cpp
class Solution {
public:
    string shortestCommonSupersequence(string str1, string str2) {
        int m=str1.length();
        int n=str2.length();
        vector<vector<int>> dp(m+1,vector<int>(n+1,0));
        for(int i=1;i<=m;i++){
            for(int j=1;j<=n;j++){
                if (str1[i-1]==str2[j-1]){
                    dp[i][j]=1+dp[i-1][j-1];
                }
                else{
                    dp[i][j]=max(dp[i-1][j],dp[i][j-1]);
                }
            }
        }
        int i=m;
        int j=n;
        vector<char> ans;
        while (i>=1 and j>=1){
            if(str1[i-1]==str2[j-1]){
                ans.push_back(str1[i-1]);
                i--;
                j--;
            }
            else{
                if(dp[i-1][j]>dp[i][j-1]){
                    ans.push_back(str1[i-1]);
                    i--;
                }
                else{
                    ans.push_back(str2[j-1]);
                    j--;
                }
            }
        }
        while(j>=1){
            ans.push_back(str2[j-1]);
            j--;
        }
        while(i>=1){
            ans.push_back(str1[i-1]);
            i--;
        }
        // for(auto x:ans){
        //     cout<<x<<" ";
        // }
        string res(ans.rbegin(),ans.rend());
        //string res=str{ans.rbegin(),ans.rend()};
        return res;
    
        
    }
};